pilot/pkg/serviceregistry/kube/controller/multicluster.go

```diff
@@ -377,8 +392,91 @@ func (m *Multicluster) InitSecretController(stop <-chan struct{}) {
 	m.secretController = secretcontroller.StartSecretController(
 		m.client, m.AddMemberCluster, m.UpdateMemberCluster, m.DeleteMemberCluster,
 		m.secretNamespace, m.syncInterval, stop)
+	go m.sync()
+}
+
+func (m *Multicluster) sync() {
+	defer func() {
+		m.synced.Store(true)
+	}()
+
+	// stop cancels sync wait
+	stop := make(chan struct{})
+	defer close(stop)
+
+	// timeout places an upper bound on how long sync errors can clear
+	timeout := time.After(readinessTimeout)
+
+	// wait for either readiness, or a timeout
+	done := make(chan struct{})
+	go func() {
+		m.waitForRegistriesSynced(stop)
+		close(done)
+	}()
+	select {
+	case <-done:
+		return
+	case <-timeout:
+	}
+
+	// timeout has expired, start watching for any sync error
+	err := make(chan error)
+	defer close(err)
+	go func() {
+		err <- m.waitForSyncErrors(stop)
+	}()
+
+	select {
+	case <-done:
+		return
+	case err := <-err:
+		log.Warnf("sync errors for multicluster registries still present after %v: %v", readinessTimeout, err)
+	}
+}
+
+func (m *Multicluster) waitForRegistriesSynced(stop chan struct{}) {
+	// wait for the secret controller to init registries
+	cache.WaitForCacheSync(stop, m.secretController.HasSynced)
+
+	// collect all the service registry HasSynced
+	syncs := make([]cache.InformerSynced, 0, len(m.remoteKubeControllers))
+	m.m.Lock()
+	for _, controller := range m.remoteKubeControllers {
+		syncs = append(syncs, controller.HasSynced)
+		if controller.workloadEntryStore != nil {
+			syncs = append(syncs, controller.workloadEntryStore.HasSynced)
+		}
+	}
+	m.m.Unlock()
+
+	cache.WaitForCacheSync(stop, m.secretController.HasSynced)
+}
+
+func (m *Multicluster) waitForSyncErrors(stop chan struct{}) error {
+	syncs := make([]func() error, 0, len(m.remoteKubeControllers))
+	m.m.Lock()
+	for _, controller := range m.remoteKubeControllers {
+		syncs = append(syncs, controller.SyncError)
+		if controller.configStore != nil {
+			syncs = append(syncs, controller.configStore.SyncErr)
```

Comment:
The config store SyncErr is a lot more expensive – it calls a lister, whereas the kube controllers' SyncErr impl has the error precomputed based on what's going on in HasSynced...

Reply:
What is `SyncErr` used for?